### PR TITLE
NRT clarifications

### PR DESCRIPTION
**Pull request by gewarren**
- Clarifies what "nullable reference types" refers to
- Adds a table to compare contexts

[Preview](https://review.docs.microsoft.com/en-us/dotnet/csharp/nullable-references?branch=pr-en-us-26840)

### CODE REVIEW

File: docs/csharp/nullable-references.md [hunk 1]
```diff
@@ -162,18 +164,27 @@ You must explicitly opt in to use these features in your projects. That provides
...
   - All other reference type variables are non-nullable reference types.
   - You can use the null forgiving operator, `!`, but it has no effect.
 
+The nullable annotation context and nullable warning context can be set for a project using the [`<Nullable>` element](language-reference/compiler-options/language.md) in your *.csproj* file. This element configures how the compiler interprets the nullability of types and what warnings are generated. The following table shows the allowable values and summarizes the contexts they specify.
+
+| Context | Dereference warnings | Reference types | `?` suffix | `!` operator |
+| - | - | - | - | - |
+| `disabled` | Disabled | All are nullable | Can't be used | Has no effect |
+| `enabled` | Enabled | Non-nullable unless declared with `?` | Declares nullable type | Suppresses warnings for possible `null` assignment |
+| `warnings` | Enabled | All are nullable, but members are considered *not null* at opening brace of methods | Produces a warning |  Suppresses warnings for possible `null` assignment |
```
Comment (gewarren) [inline docs/csharp/nullable-references.md:185]: This row likely needs to be modified. I was confused that using `?` produces a warning, but then it says that members are considered not-null unless declared with `?`.

File: docs/csharp/nullable-references.md [hunk 2]
```diff
@@ -151,9 +153,9 @@ The new features that protect against throwing a <xref:System.NullReferenceExcep
 - The meaning of the `class` constraint in generics changed to mean a non-nullable reference type.
 - New warnings are generated because of these new rules.
 
-You must explicitly opt in to use these features in your projects. That provides a migration path and preserves backwards compatibility. Nullable contexts enable fine-grained control for how the compiler interprets reference type variables. The **nullable annotation context** determines the compiler's behavior. There are four values for the **nullable annotation context**:
+You must explicitly opt in to use these features in your existing projects. That provides a migration path and preserves backwards compatibility. Nullable contexts enable fine-grained control for how the compiler interprets reference type variables. The **nullable annotation context** determines the compiler's behavior. There are four values for the **nullable annotation context**:
```
Comment (IEvangelist) [inline docs/csharp/nullable-references.md:156]:
```suggestion
You must explicitly opt-in to use these features in your existing projects. That provides a migration path and preserves backwards compatibility. Nullable contexts enable fine-grained control for how the compiler interprets reference type variables. The **nullable annotation context** determines the compiler's behavior. There are four values for the **nullable annotation context**:
```

File: docs/csharp/nullable-references.md [hunk 3]
```diff
@@ -162,18 +164,27 @@ You must explicitly opt in to use these features in your projects. That provides
...
   - All reference type variables are allowed to be null. However, members have the *null-state* of *not-null* at the opening brace of all methods unless declared with the `?` suffix.
   - You can use the null forgiving operator, `!`.
-- *annotations*: The compiler doesn't perform null analysis or emits warnings when code might dereference `null`.
+- *annotations*: The compiler doesn't perform null analysis or emit warnings when code might dereference `null`.
   - All new nullable warnings are disabled.
   - You can use the `?` suffix to declare a nullable reference type.
   - All other reference type variables are non-nullable reference types.
   - You can use the null forgiving operator, `!`, but it has no effect.
 
+The nullable annotation context and nullable warning context can be set for a project using the [`<Nullable>` element](language-reference/compiler-options/language.md) in your *.csproj* file. This element configures how the compiler interprets the nullability of types and what warnings are generated. The following table shows the allowable values and summarizes the contexts they specify.
```
Comment (IEvangelist) [inline docs/csharp/nullable-references.md:179]:
```suggestion
The nullable annotation context and nullable warning context can be set for a project using the [`<Nullable>` element](language-reference/compiler-options/language.md) in your *.csproj* file. This element configures how the compiler interprets the nullability of types and what warnings are emitted. The following table shows the allowable values and summarizes the contexts they specify.
```

The "emit" word is used more often when referring to warnings from nullable contexts, for consistency this felt a bit more applicable.

**Review by IEvangelist — COMMENTED** [review]
A few nits

File: docs/csharp/nullable-references.md (hunk 1 above)
Comment (BillWagner) [reply docs/csharp/nullable-references.md:185]: I think this works. I'll make a suggestion on the entire table to add a column for "assignment to null warnings"

File: docs/csharp/nullable-references.md [hunk 4]
```diff
@@ -162,18 +164,27 @@ You must explicitly opt in to use these features in your projects. That provides
...
+
+| Context | Dereference warnings | Reference types | `?` suffix | `!` operator |
+| - | - | - | - | - |
+| `disabled` | Disabled | All are nullable | Can't be used | Has no effect |
+| `enabled` | Enabled | Non-nullable unless declared with `?` | Declares nullable type | Suppresses warnings for possible `null` assignment |
+| `warnings` | Enabled | All are nullable, but members are considered *not null* at opening brace of methods | Produces a warning |  Suppresses warnings for possible `null` assignment |
+| `annotations` | Disabled | Non-nullable unless declared with `?` | Declares nullable type | Has no effect |
```
Comment (BillWagner) [inline docs/csharp/nullable-references.md:186]: I tried adding a column for "Assignment warnings".  Does that help explain the warnings on annotations in that row?

```suggestion
| Context | Dereference warnings | Assignment warnings | Reference types | `?` suffix | `!` operator |
| - | - | - | - | - |
| `disabled` | Disabled | Disabled | All are nullable | Can't be used | Has no effect |
| `enabled` | Enabled | Enabled | Non-nullable unless declared with `?` | Declares nullable type | Suppresses warnings for possible `null` assignment |
| `warnings` | Enabled | Not applicable | All are nullable, but members are considered *not null* at opening brace of methods | Produces a warning |  Suppresses warnings for possible `null` assignment |
| `annotations` | Disabled | Disabled | Non-nullable unless declared with `?` | Declares nullable type | Has no effect |
```

**Review by BillWagner — APPROVED** [review]
This is a great set of enhancements @gewarren 

I had one suggestion in addition to David's. I'm not sure if the extra column in the table helps, or adds confusion. Let me know what you think.

File: docs/csharp/nullable-references.md (hunk 2 above)
Comment (gewarren) [reply docs/csharp/nullable-references.md:156]: I'm going to leave this as-is based on https://www.merriam-webster.com/dictionary/opt%20in.

**Comment by gewarren** [comment]
> This is a great set of enhancements @gewarren
> 
> I had one suggestion in addition to David's. I'm not sure if the extra column in the table helps, or adds confusion. Let me know what you think.

I like it - thanks.

**Pull request merged**